### ISSUE TITLE
Problem to set the scorer function via Python wrapper

### ISSUE THREAD

**Issue by CAClaveau** [lacava/feat#219]
Hi,

Firstly, thank you for sharing your work. I learned a lot about GP.

I'm currently using FEAT *Python wrapper* for my thesis.

I had some issues to set the scorer to "zero_one" metric (with "CART" ml for classification). I managed to solve my problem : this was because the scorer string was passed inside the curly braces of "feat.cc" constructor rather than in the initialization list of the constructor, such that "(...), print_pop(print_pop), scorer(scorer)". Otherwise, the scorer string was always initialized to its default value, namely an empty string, and then set to "log" scorer after "params.set_scorer" was called, as we can see when "msg("scorer set to " + scorer,3)" is printed.
By the way, I'm not sure using "ml.compare("LR") || ml.compare("SVM")" in "params.set_scorer" does what you expect because "compare" method returns an integer, and so this Boolean expression will never give you "False".

A few other questions / comments:

At this time, RandomForest ml doesn't work because "feature_importances" function is not yet implemented, is it ?

Why "bal_zero_one" metric is not compatible with 2-class classification problem ? I needed to change
"SGVector<double> _tmp = dynamic_pointer_cast<sh::CMulticlassLabels>(labels)->get_labels()" for
"SGVector<double> _tmp = dynamic_pointer_cast<sh::CBinaryLabels>(labels)->get_labels()" to make it work.

Just to be sure, the survival step using nsga2 method is performed based on the individuals ranking done according to their score assessed on the hold out validation set ?

Thank you,
CA.

**Comment by lacava** [comment]
> I'm currently using FEAT Python wrapper for my thesis.

Great!

> I had some issues to set the scorer to "zero_one" metric (with "CART" ml for classification). I managed to solve my problem : this was because the scorer string was passed inside the curly braces of "feat.cc" constructor rather than in the initialization list of the constructor, such that "(...), print_pop(print_pop), scorer(scorer)". Otherwise, the scorer string was always initialized to its default value, namely an empty string, and then set to "log" scorer after "params.set_scorer" was called, as we can see when "msg("scorer set to " + scorer,3)" is printed.

I think we just fixed this on branch https://github.com/lacava/feat/tree/scorerfix

> At this time, RandomForest ml doesn't work because "feature_importances" function is not yet implemented, is it ?

Yes, we need to implement importance scores to get RF to work. They are implemented in CART, so we may be able to use MyCARTree to make this happen. help is welcome.

> Why "bal_zero_one" metric is not compatible with 2-class classification problem ? I needed to change
"SGVector _tmp = dynamic_pointer_castsh::CMulticlassLabels(labels)->get_labels()" for
"SGVector _tmp = dynamic_pointer_castsh::CBinaryLabels(labels)->get_labels()" to make it work.

Can you explain more where this is happening?

> Just to be sure, the survival step using nsga2 method is performed based on the individuals ranking done according to their score assessed on the hold out validation set ?

Selection and survival always use training data. Validation is only used at the end to select the final model, as it should only be used once.

**Comment by lacava** [comment]
> At this time, RandomForest ml doesn't work because "feature_importances" function is not yet implemented, is it ?

as of PR #224 , RandomForest is supported.

**Comment by lacava** [comment]
@CAClaveau, I'm closing this issue now and refer you to issue #226 for the bal_zero_one loss on binary classification problems. The other two issues you raised (random forest support and changing the scorer) have been addressed.
